### PR TITLE
Add MemoryConsumption::memory_consumption() overloads for Lazy<T>

### PR DESCRIPTION
**Pull request by tamiko**
- base/memory_consumption.h: add specialization for std::optional<T>
- base/lazy.h: add memory_consumption method to Lazy<T>

In reference to #16204

### CODE REVIEW

File: include/deal.II/base/lazy.h [hunk 1]
```diff
@@ -396,5 +403,14 @@ Lazy<T>::value_or_initialize(const Callable &creator)
 }
 
 
+template <typename T>
+std::size_t
+Lazy<T>::memory_consumption() const
+{
+  return MemoryConsumption::memory_consumption(object) +
+         sizeof(object_is_initialized) + sizeof(initialization_mutex);
```
Comment (kronbichler) [inline include/deal.II/base/lazy.h:411]: Would
```suggestion
  return MemoryConsumption::memory_consumption(object) +
         sizeof(*this) - sizeof(object);
```
be more precise? Not that it is going to be a large difference, but due to alignment reasons, I expect that `std::atomic<bool>` is of size 1, but will likely require 8 or 16 bytes, depending on the alignment of the mutex.

**Review by kronbichler — APPROVED** [review]
OK apart from a minor detail (which you might choose to ignore).

File: include/deal.II/base/lazy.h (hunk 1 above)
Comment (tamiko) [reply include/deal.II/base/lazy.h]: Agreed.

**Comment by bangerth** [comment]
You've got to re-indent...

**Review by kronbichler — APPROVED** [review]

**Pull request merged**